modin/data_management/functions/groupby_function.py

```diff
@@ -20,174 +20,243 @@
+        )
+
+    @classmethod
+    def map(
+        cls,
+        df,
+        other=None,
+        axis=0,
+        by=None,
+        groupby_args=None,
+        map_func=None,
+        map_args=None,
+        drop=False,
+    ):
+        # Set `as_index` to True to track the metadata of the grouping object
+        # It is used to make sure that between phases we are constructing the
+        # right index and placing columns in the correct order.
+        groupby_args["as_index"] = True
+        groupby_args["observed"] = True
+        if other is not None:
+            other = other.squeeze(axis=axis ^ 1)
+            if isinstance(other, pandas.DataFrame):
+                df = pandas.concat(
+                    [df] + [other[[o for o in other if o not in df]]],
+                    axis=1,
                 )
-            assert axis == 0, "Can only groupby reduce with axis=0"
+                other = list(other.columns)
+            by_part = other
+        else:
+            by_part = by
+
+        apply_func = cls.try_filter_dict(map_func, df)
+        result = apply_func(
+            df.groupby(by=by_part, axis=axis, **groupby_args), **map_args
+        )
+        return result
+
+    @classmethod
+    def reduce(
+        cls,
+        df,
+        partition_idx=0,
+        axis=0,
+        groupby_args=None,
+        reduce_func=None,
+        reduce_args=None,
+        drop=False,
+        **kwargs,
+    ):
+        by_part = list(df.index.names)
+        if drop and len(df.columns.intersection(by_part)) > 0:
+            df.drop(columns=by_part, errors="ignore", inplace=True)
+
+        groupby_args = groupby_args.copy()
+        method = kwargs.get("method", None)
+        as_index = groupby_args["as_index"]
 
-            if numeric_only:
-                qc = query_compiler.getitem_column_array(
-                    query_compiler._modin_frame._numeric_columns(True)
+        # Set `as_index` to True to track the metadata of the grouping object
+        groupby_args["as_index"] = True
+
+        # since now index levels contain out 'by', in the reduce phace
+        # we want to group on these levels
+        groupby_args["level"] = list(range(len(df.index.names)))
+
+        apply_func = cls.try_filter_dict(reduce_func, df)
+        result = apply_func(df.groupby(axis=axis, **groupby_args), **reduce_args)
+
+        if not as_index:
+            insert_levels = partition_idx == 0 and (drop or method == "size")
+            result.reset_index(drop=not insert_levels, inplace=True)
+
+        return result
+
+    @classmethod
+    def caller(
+        cls,
+        query_compiler,
+        by,
+        axis,
+        groupby_args,
+        map_args,
+        map_func,
+        numeric_only=True,
+        **kwargs,
+    ):
+        if not isinstance(by, (type(query_compiler), str)):
+            by = try_cast_to_pandas(by, squeeze=True)
+            default_func = (
+                (lambda grp: grp.agg(map_func))
```

Comment:
Why additional () are needed?

Reply:
because otherwise that "if" expression will be treated like it is inside lambda
```python
>>> result = (
...     lambda *args: 42
...     if False
...     else max
... )
>>> result()  # `result` variable contains lambda that returns `max` function
<built-in function max>
>>>
>>> result = (
...     (lambda *args: 42)
...     if False
...     else max
... )
>>> result()  # got 'max' function in the `result` variable as expected
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: max expected 1 arguments, got 0
```